Use recommended pattern in testing example

htunnicliff opened the pull request:
Since the official linter for React Testing Library (`eslint-plugin-testing-library`) [recommends](https://github.com/testing-library/eslint-plugin-testing-library/blob/main/docs/rules/prefer-screen-queries.md) using `screen` to write queries, this MR updates the Testing Library example to follow the pattern recommended by the linter.

> DOM Testing Library (and other Testing Library frameworks built on top of it) exports a `screen` object which has every query (and a debug method). This works better with autocomplete and makes each test a little simpler to write and maintain.

> This rule aims to force writing tests using built-in queries directly from `screen` object rather than destructuring them from `render` result. Given the screen component does not expose utility methods such as `rerender()` or the `container` property, it is correct to use the `render` returned value in those scenarios.

See the `prefer-screen-queries` rules docs for more info: https://github.com/testing-library/eslint-plugin-testing-library/blob/main/docs/rules/prefer-screen-queries.md

## Documentation / Examples

- [ ] Make sure the linting passes

htunnicliff commented:
@leerob will do!

Are there any plans to incorporate the recommended linter(s) for Testing Library and Cypress, either as standalone dependencies in the examples or within next/lint itself?

leerob commented:
I like the idea of extending the default ESLint setup on the example to include this package! 👍

htunnicliff commented:
@leerob I added linting configuration and updated the example test. For clarity, I also swapped in the name "Home" for the tests, since the example is not testing the App but rather the `Home` component. I also updated the name of testing-library.js to `index.test.jsx`, which more closely matches real-world test file names.

delbaoliveira commented:
Hey @htunnicliff, thank you for submitting this PR and updating the example! 💪🏼

delbaoliveira commented on docs/testing.md:
```diff
@@ -179,6 +179,7 @@ module.exports = {
     '^.+\\.(jpg|jpeg|png|gif|webp|svg)$': '<rootDir>/__mocks__/fileMock.js',
   },
   testPathIgnorePatterns: ['<rootDir>/node_modules/', '<rootDir>/.next/'],
+  testEnvironment: 'jsdom',
```
Not a major issue. What do you think about setting the environment locally (per file) instead of globally as [recommended by Jest](https://jestjs.io/blog/2021/05/25/jest-27)?

htunnicliff replied:
I've gone back and forth on that recommended pattern myself in real life projects 😂 

I think either way could work, but I'd be happy to modify this to follow the JSDoc approach if it is this project's preference?

delbaoliveira replied:
Ok 👍  Let's go with per file since not every test will require `jsdom` and follows Jest recommendation. I added a note in [docs here](https://github.com/vercel/next.js/pull/28209/files#diff-c22885d97fa21ad974a3f5f982d92b6f45edc251fd34bc14223492ebd7391245R306s) so people can set it globally if they want to.

htunnicliff replied:
@delbaoliveira I made the changes. I think there may be merge conflicts with the changes in #28209 though.

delbaoliveira replied:
Thank you and yes, there will be some merge conflicts. That's ok though, I can update the other pull request! 🙏🏼

htunnicliff commented:
@delbaoliveira from my end it looks like this PR is still marked as having "requested changes." Is there anything else I should do to help get this merged?

leerob reviewed (approved):
Thank you so much!

leerob commented:
Looking like linting is failing here, which is why it didn't auto merge.

htunnicliff commented:
> Looking like linting is failing here, which is why it didn't auto merge.

@leerob it looks like there is a conflict between with-jest's `.eslintrc.json` and the repo's `.eslintrc.json`. 

Looking at other example folders, I see that `root: true` is set, so I added that in 497d698 – hopefully, it fixes the lint errors 🤞

htunnicliff commented:
It looks like linting is still failing even when set to `root`. 

Does the configuration of this monorepo require that all of the dependencies in `examples/with-jest/package.json` also be added to `/package.json`?

Aside from that, I'm not sure what else to try at this point. 🤔

ijjk commented:
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | htunnicliff/next.js patch-1 | Change |
| - | - | - | - |
| buildDuration | 15.8s | 15.9s | ⚠️ +190ms |
| buildDurationCached | 3.6s | 3.7s | ⚠️ +68ms |
| nodeModulesSize | 61.7 MB | 61.7 MB | ✓ |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | htunnicliff/next.js patch-1 | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.844 | 2.795 | -0.05 |
| / avg req/sec | 879.14 | 894.46 | +15.32 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.578 | 1.599 | ⚠️ +0.02 |
| /error-in-render avg req/sec | 1584.49 | 1563.85 | ⚠️ -20.64 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | htunnicliff/next.js patch-1 | Change |
| - | - | - | - |
| 745.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.2 kB | 42.2 kB | ✓ |
| main-HASH.js gzip | 23.3 kB | 23.3 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 67.2 kB | 67.2 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | htunnicliff/next.js patch-1 | Change |
| - | - | - | - |
| polyfills-a4..dd70.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | htunnicliff/next.js patch-1 | Change |
| - | - | - | - |
| _app-HASH.js gzip | 979 B | 979 B | ✓ |
| _error-HASH.js gzip | 194 B | 194 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 329 B | 329 B | ✓ |
| dynamic-HASH.js gzip | 2.67 kB | 2.67 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 918 B | 918 B | ✓ |
| image-HASH.js gzip | 4.14 kB | 4.14 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 1.66 kB | 1.66 kB | ✓ |
| routerDirect..HASH.js gzip | 318 B | 318 B | ✓ |
| script-HASH.js gzip | 387 B | 387 B | ✓ |
| withRouter-HASH.js gzip | 320 B | 320 B | ✓ |
| bb14e60e810b..30f.css gzip | 125 B | 125 B | ✓ |
| Overall change | 13 kB | 13 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | htunnicliff/next.js patch-1 | Change |
| - | - | - | - |
| _buildManifest.js gzip | 491 B | 491 B | ✓ |
| Overall change | 491 B | 491 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | htunnicliff/next.js patch-1 | Change |
| - | - | - | - |
| index.html gzip | 540 B | 540 B | ✓ |
| link.html gzip | 553 B | 553 B | ✓ |
| withRouter.html gzip | 534 B | 534 B | ✓ |
| Overall change | 1.63 kB | 1.63 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Webpack 4 Mode</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | htunnicliff/next.js patch-1 | Change |
| - | - | - | - |
| buildDuration | 13.3s | 13.3s | -24ms |
| buildDurationCached | 5.8s | 5.5s | -271ms |
| nodeModulesSize | 61.7 MB | 61.7 MB | ✓ |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | htunnicliff/next.js patch-1 | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.879 | 2.898 | ⚠️ +0.02 |
| / avg req/sec | 868.27 | 862.81 | ⚠️ -5.46 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.611 | 1.647 | ⚠️ +0.04 |
| /error-in-render avg req/sec | 1552.18 | 1517.6 | ⚠️ -34.58 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | htunnicliff/next.js patch-1 | Change |
| - | - | - | - |
| 16.HASH.js gzip | 186 B | 186 B | ✓ |
| 677f882d2ed8..HASH.js gzip | 14.1 kB | 14.1 kB | ✓ |
| framework.HASH.js gzip | 41.9 kB | 41.9 kB | ✓ |
| main-HASH.js gzip | 10.7 kB | 10.7 kB | ✓ |
| webpack-HASH.js gzip | 1.19 kB | 1.19 kB | ✓ |
| Overall change | 68.1 kB | 68.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | htunnicliff/next.js patch-1 | Change |
| - | - | - | - |
| polyfills-a4..dd70.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | htunnicliff/next.js patch-1 | Change |
| - | - | - | - |
| _app-HASH.js gzip | 964 B | 964 B | ✓ |
| _error-HASH.js gzip | 3.8 kB | 3.8 kB | ✓ |
| amp-HASH.js gzip | 552 B | 552 B | ✓ |
| css-HASH.js gzip | 333 B | 333 B | ✓ |
| dynamic-HASH.js gzip | 2.87 kB | 2.87 kB | ✓ |
| head-HASH.js gzip | 3.06 kB | 3.06 kB | ✓ |
| hooks-HASH.js gzip | 924 B | 924 B | ✓ |
| index-HASH.js gzip | 231 B | 231 B | ✓ |
| link-HASH.js gzip | 1.64 kB | 1.64 kB | ✓ |
| routerDirect..HASH.js gzip | 298 B | 298 B | ✓ |
| script-HASH.js gzip | 3.03 kB | 3.03 kB | ✓ |
| withRouter-HASH.js gzip | 295 B | 295 B | ✓ |
| 30809af5c834..565.css gzip | 125 B | 125 B | ✓ |
| Overall change | 18.1 kB | 18.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | htunnicliff/next.js patch-1 | Change |
| - | - | - | - |
| _buildManifest.js gzip | 500 B | 500 B | ✓ |
| Overall change | 500 B | 500 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | htunnicliff/next.js patch-1 | Change |
| - | - | - | - |
| index.html gzip | 584 B | 584 B | ✓ |
| link.html gzip | 596 B | 596 B | ✓ |
| withRouter.html gzip | 577 B | 577 B | ✓ |
| Overall change | 1.76 kB | 1.76 kB | ✓ |

</details>

</details>
Commit: 8bb1152959578f7dc518c24036b20fa9b33223bd
